### DISCUSSION TITLE
enhance the camera tutorial

### DISCUSSION THREAD

**Discussion by HLFILIOZAT** [pyvista/pyvista#5031] [Q&A]
Hello there !

I would like to share with you some enhancement of the pyvista camera tutorial.
My first aim was to make it dynamic and interactive.
A secondary objective has quickly appeared to explain / understand the links between the actions the user can do on the mouse (and/or the keyboard) and the camera movements and updates.

So, the plot is still splitted in two halves:
- the lower one (with the grid) to show the object (the bunny, the dragon or the airplane...) and get the user interactions...
- the upper one to display the object and the camera as an additional scene object, with the frustum (near and far plane), the focal point, etc.

I have changed the original 1.0 frustum view to a 16/9 format for better understanding.

Here are some limitations that I have found:
Any advice to make them disappear and/or improve the code will be strongly appreciated !

- ### Scene refresh callback:
I do not know how to trig a continuous refresh.
I have tried to add_render_callback, but it is not activated.
So, I have only set tracking on mouse buttons click...
-> After an interactive action on the object view, the user must click on the mouse to update the scene view.

- ### Roll display:
I would like to add the display of the roll angle of the camera on the scene.
-> Not done yet !

- ### Unknown interactions:
elevation and azimuth always remain equal to 0. view_angle stays at 30.
I don't know how to modify these camera attributes by interactions.

- ### Deactivated interactions:
 up / down keys do not zoom in / out. Why ?
mouse wheel and right-button clic/drag perfectly do.

_CODE:_
```python
import pyvista as pv
from pyvista import examples
import numpy as np
print(pv.Report())
pv.set_plot_theme("dark")

#object = examples.download_bunny().scale(100) # -> view_xy
#object = examples.download_dragon().scale(100)
object = examples.load_airplane() # view yz


pl = pv.Plotter(shape=(2, 1))

pl.subplot(1, 0)
pl.add_text("Object View")
pl.add_mesh(object)
pl.show_grid()
pl.view_yz()
pl.add_camera_orientation_widget()


pl.subplot(0, 0)
pl.add_text("Scene View (object + camera)")
pl.add_mesh(object)
pl.view_xz()


def cb(point):
    #print("CLICK AT:", point) # /!\ doc non respectée !!!

    global pl
    pl.subplot(1, 0)
    #print("DEBUG CAMERA:", pl.camera.position, pl.camera.focal_point, pl.camera.clipping_range )
    camera1 = pl.camera
    
    pl.subplot(0,0)
    line = pv.Line( camera1.position, camera1.focal_point)
    pl.add_mesh(line, name='line', color="cyan")

    frustum = camera1.view_frustum(16/9)
    pl.add_mesh(frustum, name='frustum', style="wireframe", color='cyan' )

    near_range, far_range = camera1.clipping_range
    
    print( "DEBUG: CAMERA:" )
    print( "\troll=", camera1.roll)       # float
    print( "\telev=", camera1.elevation)  # float
    print( "\tazim=", camera1.azimuth)    # float
    print( "\tvang=", camera1.view_angle) # float
    print( "\t  up=", camera1.up)         # 3D vector
    print( "\t dir=", camera1.direction)  # 3D vector

    pl.add_point_labels(
        [
            camera1.position,
            camera1.position + np.array(camera1.direction) * near_range,
            camera1.position + np.array(camera1.direction) * far_range,
            camera1.focal_point,
        ],
        ["Camera Position", "Near Clipping Plane", "Far Clipping Plane", "Focal Point"],
        name='label_points',
        margin=0,
        fill_shape=False,
        font_size=14,
        # shape_color="gray",
        point_color="red",
        # text_color="white"
    )

    return


pl.track_click_position( cb, side='l', viewport=True ) # TODO replace click callback by something more smooth !
pl.track_click_position( cb, side='r', viewport=True ) # TODO mouse wheel events

#pl.add_on_render_callback( cb, render_event=True ) # TODO /!\ cb( plot ) instead of point !


#pl.unlink_views() # default
pl.show()
```

**Comment by banesullivan** [comment]
Thanks for sharing, @HLFILIOZAT!!

**Comment by HLFILIOZAT** [comment]
I have added arrows at the camera position, to show camera vectors and XYZ axes.
It is very useful to understand how the mouse movements affect the camera position and angles.
![image](https://github.com/pyvista/pyvista/assets/145444197/f7f4852f-01dc-445e-b7b7-adae3d635a9b)

```python
    fact = camera1.distance/np.linalg.norm(camera1.direction)
    for i,d in enumerate( [
         (camera1.direction, 'cyan',   fact/10 ),
         (camera1.up,        'yellow', fact/10 ),
         ([1,0,0],           'red',    fact/20 ),
         ([0,1,0],           'green',  fact/20 ),
         ([0,0,1],           'blue',   fact/20 )
        ] ):
        direction, color, mag = d
        pl.add_arrows( cent=np.array( camera1.position ), 
                       direction=np.array(direction),
                       mag=mag, 
                       name='arrow%d'%i,
                       color=color )
```
